File: src/algorithms/arrays/move-zeros/sources/MoveZeros.java
// Move Zeros to End — O(n) two-pointer: write pointer tracks next write position, read pointer scans
public class MoveZeros {
    public static int[] moveZeros(int[] inputArray) {
        int[] result = inputArray.clone();
        int writePointer = 0; // @step:initialize

        for (int readPointer = 0; readPointer < result.length; readPointer++) {
            int currentElement = result[readPointer]; // @step:compare
            if (currentElement != 0) { // @step:compare
                int tempValue = result[writePointer]; // @step:swap
                result[writePointer] = currentElement; // @step:swap
                result[readPointer] = tempValue; // @step:swap
                writePointer++; // @step:visit
            }
        }

        return result; // @step:complete
    }
}

File: src/algorithms/arrays/two-pointer/move-zeros/sources/move-zeros.go
// Move Zeros to End — O(n) two-pointer: write pointer tracks next write position, read pointer scans
package movezeros

func moveZeros(inputArray []int) []int {
	result := make([]int, len(inputArray))
	copy(result, inputArray)
	writePointer := 0 // @step:initialize

	for readPointer := 0; readPointer < len(result); readPointer++ {
		currentElement := result[readPointer] // @step:compare
		if currentElement != 0 {             // @step:compare
			result[writePointer], result[readPointer] = result[readPointer], result[writePointer] // @step:swap
			writePointer++ // @step:visit
		}
	}

	return result // @step:complete
}
